pm: doc improvements and minor enhancements

gmarull opened the pull request:
This PR tries to improve the organization of the PM subsystem documentation (Doxygen part for now). Some minor enhancements have also been done.

Note: better review on a per-commit basis.

gmarull committed:
doc: doxygen: add group definition file

Add an external group definition file. It can be used to create high
level group for things like subsystems, as there is no header such as
"subsys.h" where these type of logical groups can be defined.

This approach is followed by projects such as pipewire to organize
Doxygen modules.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: cleanup Doxygen groups and inclusion

- Add a new `subsys_pm` group, part of `subsys`
- Improve group naming
- Include conditional code using __DOXYGEN__, it allows to have better
  control of inclusion.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: state: use code/endcode for code examples

Enclode code examples using code/endcode Doxygen commands.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: document deprecated functions and macros

Document the deprecated functions and macros, tag them with @deprecated
command to inform about the replacement.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove unused flag

PM_DEVICE_FLAG_COUNT is not used, so remove it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: adjust flag names

Some flags were using `FLAGS` instead of `FLAG`.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: hide internal flags from public docs

The PM flags are only used internally, so there is no need to include
them in the documentation.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove internal comment

The comment provides internal details which are not relevant.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: improve pm_device_busy* APIs docs

Make the API docs more concise.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: add dummy implementation for wakeup API

Add a dummy implementation for the wakeup API, so that it can still be
used when no PM_DEVICE is enabled.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: improve wakeup API docs

Improve the Doxygen docs for the pm_device_wakeup* APIs.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: hide struct pm_device from docs

This structure is meant for internal usage, same as its initializer.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove redundant initialization field

Z_PM_DEVICE_INIT is used as a static initializer, so all non-specified
fields will already be set to zero.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: add missing return code to pm_device_state_get

-ENOSYS can be returned if the device does not support PM.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: state: make helper macros internal

Some macros were helpers for the state initializers (e.g. to assist
UTIL_LISTIFY). In order to make the API more clear, they have been
prefixed with Z_ to emphasize they are internal and they have also been
removed from documentation.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: doxyfile: remove redundant predefinitions

PM related pre-definitions are not required, so remove them.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin reviewed (changes requested):
Other than make the device state macros private. This pr looks good for me.

gmarull committed:
pm: device: rename INIT_PM_DEVICE_RUNTIME

Rename the macro used to initialize the runtime fields to
Z_PM_DEVICE_RUNTIME_INIT (emphasize it is internal and follows parent
macro naming).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on include/pm/state.h:
```diff
@@ -154,14 +154,16 @@ struct pm_state_info {
  * @return pm_state_info item from 'cpu-power-states' property at index 'i'
  */
-#define PM_STATE_INFO_DT_ITEM_BY_IDX(node_id, i)                    \
+#define Z_PM_STATE_INFO_DT_ITEM_BY_IDX(node_id, i)                    \
```
Adjusted, only flagged as internal `UTIL_LISTIFY` helpers.

gmarull commented on include/pm/device.h:
```diff
@@ -112,15 +112,6 @@ struct pm_device {
-#else
-#define INIT_PM_DEVICE_RUNTIME(obj)
-#endif /* CONFIG_PM_DEVICE_RUNTIME */
-
```
gone back to original method

ceolin commented on include/pm/device.h:
```diff
@@ -90,12 +93,11 @@ typedef int (*pm_device_action_cb_t)(const struct device *dev,
  */
 struct pm_device {
-#ifdef CONFIG_PM_DEVICE_RUNTIME
+#if defined(CONFIG_PM_DEVICE_RUNTIME) || defined(__DOXYGEN__)
```
why this conditional ? `defined(__DOXYGEN__)`

gmarull replied:
It can be used to unconditionally include sections of the docs without the need of adding predefines in the Doxygen side. A live example:

https://github.com/zephyrproject-rtos/zephyr/blob/811a09bd83052a40220d70eb3550048915d0be23/include/drivers/pinctrl.h#L58-L65

It allows for more fine-grained inclusion compared to using predefines.
